### PR TITLE
Output a consistent format when generating error json

### LINKED ISSUE #89387: Solidify & document guarantees about structure of error responses
What guarantees do we offer about the formatting of exception responses? Today the [docs give some examples](https://www.elastic.co/guide/en/elasticsearch/reference/current/common-options.html#common-options-error-options) but do not accurately describe the bits on which clients can rely, and our responses are [quite different](https://github.com/elastic/elasticsearch/blob/399a8ac283bb5819016c3791f2a9f40c723b9e4c/server/src/main/java/org/elasticsearch/ElasticsearchException.java#L551-L564) if the user sets `http.detailed_errors.enabled: false` which [breaks clients that are overly strict about these things](https://github.com/elastic/beats/issues/32676). It's unclear whether there are other situations that might result in differently-shaped error responses.

Some YAML tests assert things about `error.type` and `error.reason` and so on, but we don’t run those tests with detailed errors disabled (they would fail if we did). I believe there are more clients out there making decisions according to parts of the response (e.g. `error.type`) which may not always even exist, or may be unstable because they’re not covered by any tests.

It would be very helpful to clients if we could describe a general structure for error responses, and somehow guarantee that however we're configured we always follow the correct structure.

### PR DESCRIPTION
**Pull request by thecoop**
Now, error fields will always have 'type' and 'reason' fields, and the information in those fields is the same regardless of whether the output is detailed or not. This also reverses the test changes in #116330, due to the deprecation warning not (generally) being there anymore.

This fixes #89387

### CODE REVIEW

**Comment by thecoop** [comment]
Note this does have some BwC concerns, as this changes the output format when not detailed, but from #89387 we consider the current behaviour a bug

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 1]
```diff
@@ -505,12 +505,12 @@ public void testGetDetailedMessage() {
     public void testToXContent() throws IOException {
         {
             ElasticsearchException e = new ElasticsearchException("test");
-            assertExceptionAsJson(e, """
+            assertThrowableAsJson(e, """
```
Comment (grcevski) [inline server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:508]: What's the reason for changing the assert method name from exception to throwable? Throwable will cover errors too, which we shouldn't be getting, i.e they terminate Elasticsearch.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 2]
```diff
@@ -610,6 +610,85 @@ public void testToXContent() throws IOException {
         }
     }
 
+    public void testGenerateFailureToXContentWithNoDetails() throws IOException {
+        {
+            Exception ex;
+            if (randomBoolean()) {
```
Comment (grcevski) [inline server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:616]: Perhaps a better test would be to do random level of wrapping (up to 10), e.g. a loop of n times that wraps one exception to another.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 3]
```diff
@@ -610,6 +610,85 @@ public void testToXContent() throws IOException {
...
+        }
+        {
+            ParsingException ex = new ParsingException(1, 2, "foobar", null);
+            assertFailureAsJson(ex, """
+                {"error":{"type":"parsing_exception","reason":"foobar"}}""", false);
+        }
+
+        { // render header and metadata
+            ParsingException ex = new ParsingException(1, 2, "foobar", null);
+            ex.addMetadata("es.test1", "value1");
```
Comment (grcevski) [inline server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:633]: What's the significance of adding the metadata and the headers, should they be verified in the expected output if we get an exception? The comment says we render them, I'm probably not seeing how they are validated.

File: server/src/test/java/org/elasticsearch/rest/RestResponseTests.java [hunk 4]
```diff
@@ -85,19 +85,6 @@ public void testDetailedExceptionMessage() throws Exception {
             {"type":"file_not_found_exception","reason":"/foo/bar"}"""));
     }
 
-    public void testNonElasticsearchExceptionIsNotShownAsSimpleMessage() throws Exception {
```
Comment (grcevski) [inline server/src/test/java/org/elasticsearch/rest/RestResponseTests.java:88]: Why did we remove this test altogether? Do we have other tests covering what happens if non-elasticsearch exception is thrown? If the expected outcome is the same, perhaps we can refactor the above tests to demonstrate the error is consistent between throwing ES and non-ES exception.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 3 above)
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:633]: Comment was left over from where I copied the exception from - updated

File: server/src/test/java/org/elasticsearch/rest/RestResponseTests.java (hunk 4 above)
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/rest/RestResponseTests.java:88]: This tests that the error message is just the string "No ElasticsearchException found", which is no longer the behaviour of this class. Checking the error messages when detailed is off is now covered by `testGenerateFailureToXContentWithNoDetails`, rendering this test superfluous.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 2 above)
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:616]: Good idea

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 1 above)
Comment (grcevski) [reply server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:508]: Sorry to be nitpicking here, it's really a minor point, but the argument of the method is an Exception, not a Throwable. If anything shouldn't both of them be called Exception?

**Review by ChrisHegarty — CHANGES_REQUESTED** [review]
Please await the outcome of a decision regarding whether this PR is truly breaking or not.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 1 above)
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:508]: `generateThrowableXContent` takes a `Throwable`, so I can change the assert method to also take a `Throwable`, so everything matches.

**Comment by DaveCTurner** [comment]
I believe we should indeed count this as a breaking change. We may have users running with `http.detailed_errors.enabled: false` and a client that can handle the resulting (abbreviated) error responses, and such users would find their system stops working with this change.

File: server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 1 above)
Comment (grcevski) [reply server/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java:508]: OK, yeah that seems the most consistent.

**Comment by grcevski** [comment]
Should we then make this change such that we return both the new and the legacy responses? How do we typically communicate this back to the end users, telling them the legacy response field will go away?

**Comment by thecoop** [comment]
Unfortunately the two formats are mutually incompatible - in the default format, `errors` is an object, in the non-detailed format, `errors` is a string.

**Comment by DaveCTurner** [comment]
> How do we typically communicate this back to the end users

Preferably we'd deprecate the old behaviour pending its removal, and introduce some new behaviour to replace it.

In this case we could for instance deprecate setting `http.detailed_errors.enabled: false` and introduce a third option, say, `http.detailed_errors.enabled: no_cause` to avoid the deprecation warning and adopt the future behaviour today. Then in 9.0 we'd be ok to change the (deprecated) behaviour of `http.detailed_errors.enabled: false`. Since `no_cause` would do the same thing, we'd deprecate this value in 9.0 in favour of `false` and then remove it in 10.0.

**Review by grcevski — APPROVED** [review]
I don't think we've documented and made any guarantees on the output so far, when the setting is false. I think we should make this change and document the new behaviour, showing that without detailed we have the same structure except it's not showing the stack trace.

**Comment by thecoop** [comment]
BwC committee has agreed to JFDI the change in the next major release (v9)

**Comment by thecoop** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/rest/RestResponseTests.java [hunk 5]
```diff
@@ -114,11 +113,10 @@ public void testSimpleExceptionMessage() throws Exception {
         Exception t = new ElasticsearchException("an error occurred reading data", new FileNotFoundException("/foo/bar"));
         RestResponse response = new RestResponse(channel, t);
         String text = response.content().utf8ToString();
-        assertThat(text, containsString("ElasticsearchException[an error occurred reading data]"));
+        assertThat(text, containsString("an error occurred reading data"));
```
Comment (ldematte) [inline server/src/test/java/org/elasticsearch/rest/RestResponseTests.java:116]: Nit: wouldn't it better if this test is changed to be more similar to testDetailedExceptionMessage, asserting we have "type" and "reason" fields?

**Review by ldematte — APPROVED** [review]
One minor nit, but it looks good to me!

File: server/src/test/java/org/elasticsearch/rest/RestResponseTests.java (hunk 5 above)
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/rest/RestResponseTests.java:116]: Yes, good spot

**Comment by thecoop** [comment]
@elasticmachine rerun elasticsearch-ci/part-1

**Comment by thecoop** [comment]
@elasticmachine update branch

**Review by mosche — APPROVED** [review]
LGTM, just one optional thought / questions

File: server/src/main/java/org/elasticsearch/rest/RestResponse.java [hunk 6]
```diff
@@ -146,12 +147,12 @@ public RestResponse(RestChannel channel, RestStatus status, Exception e) throws
             params = new ToXContent.DelegatingMapParams(singletonMap(REST_EXCEPTION_SKIP_STACK_TRACE, "false"), params);
         }
 
-        if (channel.detailedErrorsEnabled() == false) {
+        if (channel.request().getRestApiVersion() == RestApiVersion.V_8 && channel.detailedErrorsEnabled() == false) {
             deprecationLogger.warn(
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/rest/RestResponse.java:151]: I don't think so, there's (probably) nothing directly for users to do; we consider the existing behaviour a bug, as it's surprising the format changes with non-detailed errors.

**Comment by samxbr** [comment]
Friendly reminder that this PR seems to be a breaking change for 9.0, but is missing from the 9.0 release note. We may want to add an entry to the breaking change section of [9.0 release note](https://github.com/elastic/elasticsearch/blob/cf536f894756f25c7ae874b4211a6b00372b39a1/docs/reference/release-notes/9.0.0.asciidoc#L8).

**Comment by thecoop** [comment]
See https://github.com/elastic/elasticsearch/pull/125485

**Comment by thecoop** [comment]
No, this option is not relevant for serverless

**Pull request merged**